commands/service_deploy.go

```diff
@@ -38,21 +41,32 @@ func (c *serviceDeployCmd) preRunE(cmd *cobra.Command, args []string) error {
 }
 
 func (c *serviceDeployCmd) runE(cmd *cobra.Command, args []string) error {
-	var (
-		id    string
-		valid bool
-		err   error
-	)
-	pretty.Progress("Deploying the service...", func() {
-		id, valid, err = c.e.ServiceDeploy(c.path)
-	})
+	statuses := make(chan provider.DeployStatus, 0)
+	go printDeployStatuses(statuses)
+	id, validationError, err := c.e.ServiceDeploy(c.path, statuses)
 	if err != nil {
 		return err
 	}
-	if !valid {
-		return errors.New("Service is invalid. To get more information, run: mesg-core service validate")
+	if validationError != nil {
+		pretty.DestroySpinner()
```

Comment:
Here just wonder, shouldn't deployer return `DONE` DeployedStatus on an error, it's kind of done but with error.

Because Here stoping spinner that hasn't been started in the same function looks suspicious.

The same situation is in service_dev.go

Reply:
i had the similar thing in my mind but i wasn't sure to implement this because this is not an exception to validation error. an error may occour at any time during deployment and because of this we may need to destroy spinner unexpectedly. actually, I should move this line to out of _validationError_'s nil check to make sure we call it all the time, thanks for the heads up.

Reply:
What about sending DONE to deploy statuses?

Reply:
i think errors should be only returned from funcs otherwise we'll have duplication of errors in the place where they're handled. feel free to open an issue about this so we can discuss there.